File: src/main/java/com/mybestcode/moyklassmcpserver/model/Class.java
package com.mybestcode.moyklassmcpserver.model;

import java.time.OffsetDateTime;
import java.util.List;

public record Class(
        Long id,
        String name,
        OffsetDateTime beginDate,
        Integer maxStudents,
        Status status,
        OffsetDateTime createdAt,
        Long courseId,
        PayType payType,
        Integer filialId,
        Boolean payPass,
        PayPassRules payPassRules,
        Double price,
        String priceComment,
        Boolean showDates,
        String priceForWidget,
        String color,
        List<Integer> managerIds,
        String comment,
        ClassStats stats,
        LessonSettings lessonSettings,
        WorkOff workOff,
        Invoices invoices,
        BonusProgram bonusProgram
) {
    public enum Status {
        opened, closed, archive
    }

    public enum PayType {
        full, lessons
    }

    public record PayPassRules(
            Boolean payPassReason,
            Double payPassReasonRate,
            Boolean payPassNoReason,
            Double payPassNoReasonRate
    ) {}

    public record ClassStats(
            Double income,
            Double debt
    ) {}

    public record LessonSettings(
            Access webinarAccess,
            Access videoAccess,
            Access lessonTaskAccess,
            Access homeTaskAccess
    ) {
        public enum Access {
            always, subExists, invoiceFullPay, invoiceHalfPay
        }
    }

    public record WorkOff(
            Boolean payPassWorkOff,
            Double payPassWorkOffRate,
            Boolean limitWorkOffCount,
            Integer maxWorkOffCount,
            Boolean limitCreateWorkOff,
            LimitCreateWorkOffType limitCreateWorkOffType,
            String limitCreateWorkOffPeriod
    ) {
        public enum LimitCreateWorkOffType {
            sub, period
        }
    }

    public record Invoices(
            Boolean autoCreate,
            CreateRule createRule,
            List<Integer> joinStateId,
            PayDateType payDateType,
            Double payDateDays,
            java.time.LocalDate payDate
    ) {
        public enum CreateRule {
            create, setStatus
        }

        public enum PayDateType {
            relative, exact
        }
    }

    public record BonusProgram(
            Boolean restrictBP,
            Double maxDiscount,
            Double accrueBonuses,
            Double sumLimit
    ) {}
}

File: src/types/class.ts
export type ClassStatus = 'opened' | 'closed' | 'archive';
export type ClassPayType = 'full' | 'lessons';
export type LessonAccess = 'always' | 'subExists' | 'invoiceFullPay' | 'invoiceHalfPay';
export type LimitCreateWorkOffType = 'sub' | 'period';
export type InvoiceCreateRule = 'create' | 'setStatus';
export type InvoicePayDateType = 'relative' | 'exact';

export interface PayPassRules {
  payPassReason?: boolean;
  payPassReasonRate?: number;
  payPassNoReason?: boolean;
  payPassNoReasonRate?: number;
}

export interface ClassStats {
  income?: number;
  debt?: number;
}

export interface LessonSettings {
  webinarAccess?: LessonAccess;
  videoAccess?: LessonAccess;
  lessonTaskAccess?: LessonAccess;
  homeTaskAccess?: LessonAccess;
}

export interface WorkOff {
  payPassWorkOff?: boolean;
  payPassWorkOffRate?: number;
  limitWorkOffCount?: boolean;
  maxWorkOffCount?: number;
  limitCreateWorkOff?: boolean;
  limitCreateWorkOffType?: LimitCreateWorkOffType;
  limitCreateWorkOffPeriod?: string;
}

export interface Invoices {
  autoCreate?: boolean;
  createRule?: InvoiceCreateRule;
  joinStateId?: number[];
  payDateType?: InvoicePayDateType;
  payDateDays?: number;
  payDate?: string;
}

export interface BonusProgram {
  restrictBP?: boolean;
  maxDiscount?: number;
  accrueBonuses?: number;
  sumLimit?: number;
}

export interface MoyKlassClass {
  id: number;
  name: string;
  beginDate?: string;
  maxStudents?: number;
  status: ClassStatus;
  createdAt?: string;
  courseId?: number;
  payType?: ClassPayType;
  filialId?: number;
  payPass?: boolean;
  payPassRules?: PayPassRules;
  price?: number;
  priceComment?: string;
  showDates?: boolean;
  priceForWidget?: string;
  color?: string;
  managerIds?: number[];
  comment?: string;
  stats?: ClassStats;
  lessonSettings?: LessonSettings;
  workOff?: WorkOff;
  invoices?: Invoices;
  bonusProgram?: BonusProgram;
}
